### PR TITLE
 Update security context for assign role to project group member

### PR DESCRIPTION
**Pull request by stonezdj**
The project list will contain all public projects, user is a member of this project, or user is in the group which is a member of this projects.
Change the behaviour of user roles, if the user is not a member of this project, then return the user's groups role of current project

### CODE REVIEW

File: src/common/security/context.go [hunk 1]
```diff
@@ -34,6 +34,8 @@ type Context interface {
 	HasWritePerm(projectIDOrName interface{}) bool
 	// HasAllPerm returns whether the user has all permissions to the project
 	HasAllPerm(projectIDOrName interface{}) bool
+	//Get current user's all project
 	GetMyProjects() ([]*models.Project, error)
+	//Get user's role in provided project
 	GetProjectRoles(projectIDOrName interface{}) []int
```
Comment (reasonerjt) [inline src/common/security/context.go]: So after the group roles are introduced this func will still return the `user role` in contrast to `combined role`?  Could you remind me why we made such decision?

File: src/common/security/local/context.go [hunk 2]
```diff
@@ -89,20 +90,24 @@ func (s *SecurityContext) HasReadPerm(projectIDOrName interface{}) bool {
 
 	roles := s.GetProjectRoles(projectIDOrName)
 
-	return len(roles) > 0
+	if len(roles) > 0 {
+		return true
+	}
+
+	//Get Group Roles
+	groupRoles := s.GetRolesByGroup(projectIDOrName)
```
Comment (ywk253100) [inline src/common/security/local/context.go]: The result of GetProjectRoles() has already contained the group roles. No need to call GetRolesByGroup() again.
The same for HasWritePerm() and HasAllPerm()

File: src/common/dao/project.go [hunk 3]
```diff
@@ -241,12 +284,31 @@ func DeleteProject(id int64) error {
...
+	o := GetOrmer()
+	sql := fmt.Sprintf(
+		`select pm.role from project_member pm 
+		left join user_group ug on pm.entity_type = 'g' and pm.entity_id = ug.id 
+		where ug.ldap_group_dn in ( %s ) and pm.project_id = ? `,
+		groupDNCondition)
+	log.Debugf("sql:%v", sql)
+	if _, err := o.Raw(sql, projectID).QueryRows(&roles); err != nil {
+		log.Warningf("Error in GetRolesByLDAPGroup, error: %v", err)
+		return nil
```
Comment (ywk253100) [inline src/common/dao/project.go]: Why don't we return the err but just log a warning message?

File: src/common/security/local/context.go [hunk 4]
```diff
@@ -191,8 +203,35 @@ func (s *SecurityContext) GetProjectRoles(projectIDOrName interface{}) []int {
...
+	user := s.user
+	project, err := s.pm.Get(projectIDOrName)
+	//No user, group or project info
+	if err != nil || project == nil || user == nil || len(user.GroupList) == 0 {
+		return roles
+	}
+	//Get role by LDAP group
+	groupDNConditions := group.GetGroupDNQueryCondition(user.GroupList)
+	roles = dao.GetRolesByLDAPGroup(project.ProjectID, groupDNConditions)
+	if err != nil {
```
Comment (ywk253100) [inline src/common/security/local/context.go]: The error checking is not necessary as dao.GetRolesByLDAPGroup() doesn't return any error.

File: src/common/security/context.go (hunk 1 above)
Comment (ywk253100) [reply src/common/security/context.go]: No. It should return all roles of the user.

File: src/common/security/local/context.go [hunk 5]
```diff
@@ -191,8 +203,35 @@ func (s *SecurityContext) GetProjectRoles(projectIDOrName interface{}) []int {
...
+		return nil
+	}
 	return roles
 }
+
+// GetMyProjects ...
+func (s *SecurityContext) GetMyProjects() ([]*models.Project, error) {
```
Comment (ywk253100) [inline src/common/security/local/context.go]: Seems the function doesn't return the projects which the user is a group member of.

File: src/ui/promgr/pmsdriver/local/local.go [hunk 6]
```diff
@@ -131,14 +131,32 @@ func (d *driver) Update(projectIDOrName interface{},
...
-	}
-
-	projects, err := dao.GetProjects(query)
-	if err != nil {
-		return nil, err
+	var total int64
+	var projects []*models.Project
+	var err error
+
+	if query != nil && query.User != nil && len(query.User.GroupList) > 0 {
```
Comment (ywk253100) [inline src/ui/promgr/pmsdriver/local/local.go]: How can we handle the request that wants all project a user is a member of(no matter user member or group member)?

File: src/common/security/local/context.go (hunk 2 above)
Comment (stonezdj) [reply src/common/security/local/context.go]: Fixed it

File: src/common/dao/project.go (hunk 3 above)
Comment (stonezdj) [reply src/common/dao/project.go]: Return error now

File: src/common/security/local/context.go (hunk 4 above)
Comment (stonezdj) [reply src/common/security/local/context.go]: Return error now

File: src/common/security/local/context.go (hunk 5 above)
Comment (stonezdj) [reply src/common/security/local/context.go]: Updated

File: src/ui/api/project.go [hunk 7]
```diff
@@ -348,7 +348,13 @@ func (p *ProjectAPI) List() {
 				}
 				projects = append(projects, pros...)
 
+				user, err := dao.GetUser(models.User{Username: p.SecurityCtx.GetUsername()})
+				if err != nil {
+					p.HandleInternalServerError(fmt.Sprintf("failed to get user: %v", err))
+					return
+				}
 				mps, err := p.ProjectMgr.List(&models.ProjectQueryParam{
+					User: user,
```
[the remark this answers is not in the document]
Comment (stonezdj) [reply src/ui/api/project.go]: Updated, call projectManager.List in SecurityContext.GetMyProjects()

File: src/common/security/local/context.go (hunk 2 above)
Comment (reasonerjt) [reply src/common/security/local/context.go]: @stonezdj So do we still need  `GetRolesByGroup`?

Comment (stonezdj) [reply src/common/security/local/context.go]: This method is used in by GetProjectRoles(), still need it

File: src/common/security/context.go [hunk 8]
```diff
@@ -24,6 +24,8 @@ type Context interface {
 	IsAuthenticated() bool
 	// GetUsername returns the username of user related to the context
 	GetUsername() string
+	//GetUserGroupList used to retrieve user group information
+	GetUserGroupList() []*models.UserGroup
```
Comment (ywk253100) [inline src/common/security/context.go]: Why we need this method?

File: src/ui/api/project.go [hunk 9]
```diff
@@ -347,10 +347,10 @@ func (p *ProjectAPI) List() {
 					return
 				}
 				projects = append(projects, pros...)
-
 				mps, err := p.ProjectMgr.List(&models.ProjectQueryParam{
 					Member: &models.MemberQuery{
-						Name: p.SecurityCtx.GetUsername(),
+						Name:      p.SecurityCtx.GetUsername(),
+						GroupList: p.SecurityCtx.GetUserGroupList(),
```
Comment (ywk253100) [inline src/ui/api/project.go]: If we call SecurityCtx.GetMyProjects() directly, we can hide the getting user group list logic inside GetMyProjects and do not need the GetUserGroupList anymore?

File: src/common/security/local/context.go (hunk 2 above)
Comment (ywk253100) [reply src/common/security/local/context.go]: No need to call GetRolesByGroup() here.

File: src/ui/api/project.go (hunk 9 above)
Comment (stonezdj) [reply src/ui/api/project.go]: change to SecurityContext.GetMyProjects()

File: src/common/security/local/context.go (hunk 2 above)
Comment (stonezdj) [reply src/common/security/local/context.go]: Removed it, issue maybe caused by code merge

File: src/common/security/context.go (hunk 8 above)
Comment (stonezdj) [reply src/common/security/context.go]: Remove it

**Pull request merged**